core/src/main/java/org/apache/iceberg/CatalogUtil.java

```diff
@@ -129,6 +129,35 @@ private static void deleteFiles(FileIO io, Set<ManifestFile> allManifests) {
         });
   }
 
+  /**
+   * Get constructor for Catalog from impl. Impl must be valid class name and have zero-arg constructor.
+   *
+   * <p>We try to get the constructor for impl and if we fail we check for common error patterns. Specifically:
+   * a ClassNotFoundException for an invalid impl String and a NoClassDefFoundError if impl fails to get initialized.
+   * Otherwise the original exception is passed up and the error is assumed to be a missing no-args constructor.
+   *
+   * @param impl catalog implementation full class name
+   * @return constructor for catalog implementation impl
+   * @throws IllegalArgumentException if no-arg constructor not found or error during initialization
+   */
+  private static DynConstructors.Ctor<Catalog> getCatalogConstructor(String impl) {
+    DynConstructors.Ctor<Catalog> ctor;
+    DynConstructors.Builder ctorBuilder = DynConstructors.builder(Catalog.class).impl(impl);
+    try {
+      ctor = ctorBuilder.buildChecked();
+    } catch (NoSuchMethodException e) {
+      Throwable originalError = ctorBuilder.problems().getOrDefault(impl, e);
+      String message = String.format("Cannot initialize Catalog, missing no-arg constructor: %s", impl);
```

Comment:
Nit: "Cannot initialize Catalog" will be printed twice if the error is neither of the one below. Although I wonder if there's a specific exception thrown if the catalog class can be found but does not have a no-arg constructor, and if so we probably could catch and print that instead, since I'd suspect the chance of misconfiguring could be much larger than only missing the no-arg constructor itself, and hence "cannot be initialized" might be a better default ...

Reply:
I suspect no-arg constructor is likely to be the most common error. 'Cannot be initialized' is related to classpath issues only.  I have updated to remove the dup message.